File: 프로그래머스/3/12927. 야근 지수/야근 지수.cpp
#include <string>
#include <vector>
#include <queue>
#include <numeric>
#include <math.h>
using namespace std;

long long solution(int n, vector<int> works) {
    long long answer = 0;
    priority_queue<int> q(works.begin(), works.end());
    int k = accumulate(works.begin(), works.end(), 1);

    while (n) {
        int z = q.top();
        q.pop();
        while (z >= q.top()) {
            z--, n--, k--;
            if (!k) return 0;
            if (!n) break;
        }
        q.push(z);
    }

    while (!q.empty()) {
        int i = q.top();
        q.pop();
        answer += pow(i, 2);
    }

    return answer;
}

File: 프로그래머스/3/12927. 야근 지수/야근 지수.java
import java.util.*;

class Solution {
    public long solution(int n, int[] works) {
        long answer = 0;
        
        PriorityQueue<Integer> q = new PriorityQueue<>(Collections.reverseOrder());
        
        for (var w : works) {
            q.offer(w);
        }
        
        for (int i = 0; i < n; i++) {
            int value = q.poll();
            if (value == 0) return 0;
            q.offer(value - 1);
        }
        
        while (!q.isEmpty()) {
            int value = q.poll();
            answer += value * value;
        }
        
        return answer;
    }
}